File: 2001-number-of-pairs-of-interchangeable-rectangles/2001-number-of-pairs-of-interchangeable-rectangles.py
##Time complexity : O(n)
##Space complexity: O(n)
class Solution:
    def interchangeableRectangles(self, rectangles: List[List[int]]) -> int:
        
        ## create a hashmap as follows
        
        count = {} # {w/h : count}
        
        res = 0
        
        for w , h in rectangles:
            
            #increament the count for ratio
            
            count[w/h] = count.get(w/h , 0) + 1
            
            #  we calculate the number of pairs that can be formed using rectangles with the same aspect ratio.
            
        for c in count.values():
            res += (c*(c-1)) // 2
            
        return res    
        
        
#solution = Solution()
#result = solution.interchangeableRectangles(rectangles)
#print(result)#        

File: 2001-number-of-pairs-of-interchangeable-rectangles/2001-number-of-pairs-of-interchangeable-rectangles.cpp
class Solution {
public:
    long long interchangeableRectangles(vector<vector<int>>& rectangles) {
        
        unordered_map<long double, int> count;
        
        long double ratio;
        
        long long res = 0;
        
        for(int i = 0; i < rectangles.size(); i++) {  // Fixed typo: changed ".ize()" to ".size()"
            ratio = (long double)(rectangles[i][0]) / (long double)(rectangles[i][1]);
            
            if(count.find(ratio) != count.end()) {
                count[ratio]++;
            }
            else {
                count[ratio] = 1;
            }
        }
        
        for(auto it : count) {
            res += (long long)(it.second) * (long long)(it.second - 1) / 2;
        }
        
        return res;  
    }
};
